dqvm/src/common.lisp

```diff
@@ -24,7 +24,7 @@
 
 (defun power-of-two-p (x)
   "Is X a power of two?"
-  (check-type x unsigned-byte)
+  (check-type x (integer 1 *))
```

Comment:
I vote for invalid input here because there is no natural number $n$ such that $2^n = 0$ and passing something that's negative would mean that the caller is broken, in which case I'd like to be informed unequivocally and immediately.

In any event, we will be dropping the requirement of passing a power of two soon.

Reply:
Yes, but there is also no natural number n such that 2^n = 3.

But I respect that case for rejecting negative numbers is stronger, and that even zero is a "grey area", and I also generally agree with the principal of rejecting bad input early and loudly.

Also, if the function is going away soon, all the more reason not to fret over it :-)

Reply:
I don't want to beat a dead horse but my point is that we have a function f : X -> Y where X is the set of natural numbers and Y is the set {0, 1}. Signaling a `type-error` here means that the user passed an argument to f that is not a member of the set X.

Reply:
True, but this only raises the question: *why* should X be "the set of natural numbers" (assuming your favorite definition of "natural numbers" does not include zero), since *we* get to pick the domain of the function.

Consider the following (contrived) example:

```lisp
(if (power-of-two-p n)
  (speedy-optimized-calculation-that-require-power-of-two n)
  (slow-and-steady-calculation-that-works-on-any-interger n))
```

Is signaling an error the right thing here if the caller passes -2?

I accept there are arguments on both sides. In general, I like small simple functional utilities like this one to be permissive about inputs so the utilities are as broadly useful as possible, only rejecting inputs that make it impossible to return any sane answer. Whereas for "business logic-y" functions doing more complex things, I prefer erring on the side of caution to catch bad input as soon as possible. This is admittedly not based on much besides personal preference though.

Reply:
Reasonable people can disagree here. To me, it seems completely clear that signaling an error is the right thing to do in the example you posed.

Reply:
Fair enough. I disagree, but respect your dedication to input validation!

Reply:
nerds

Reply:
> nerds

https://youtu.be/IRsPheErBj8